Add function for fetching in progress responses, closing #220

jamesmartherus opened the pull request:
Adds a function for downloading in progress responses. Closes #220

juliasilge commented on R/fetch_inprogress.R:
```diff
@@ -0,0 +1,169 @@
+#' }
+#'
+fetch_inprogress <- function(surveyID,
+                         last_response = deprecated(),
```
With a new function, we don't need to deprecate any arguments. No one has used it yet, so we can set up a function that only uses current parameters.

juliasilge commented on R/fetch_inprogress.R:
```diff
@@ -0,0 +1,169 @@
+#'
+#' }
+#'
+fetch_inprogress <- function(surveyID,
```
```suggestion
fetch_in_progress <- function(surveyID,
```
Let's change the function name so it doesn't squash two words together.

juliasilge commented on R/fetch_inprogress.R:
```diff
@@ -0,0 +1,169 @@
+                         add_var_labels = TRUE,
+                         col_types = NULL) {
+
+  if (lifecycle::is_present(last_response)) {
+    lifecycle::deprecate_warn("3.1.2", "fetch_inprogress(last_response = )")
+  }
```
Same note as above on deprecation

juliasilge commented on R/fetch_inprogress.R:
```diff
@@ -0,0 +1,169 @@
+
+  # See if survey already in tempdir
+  if (!force_request) {
+    if (paste0(surveyID, "_inprogress.rds") %in% list.files(tempdir())) {
```
```suggestion
    if (paste0(surveyID, "_in_progress.rds") %in% list.files(tempdir())) {
```

juliasilge replied:
Let's make this change throughout

juliasilge reviewed (changes requested):
Thank you so much for this contribution @jamesmartherus! 🙌 

One thing I notice about this PR is that it copies a fair amount of code from `fetch_survey()`. Instead of copy/pasting so much, can we create some helper functions that can be used for both functions?

The other thing we'll need here are some tests. We use vcr for testing, as [outlined in rOpenSci's book on HTTP testing](https://books.ropensci.org/http-testing/).

jamesmartherus commented:
Thanks @juliasilge! 

I've addressed the deprecation and naming comments and added some tests. Great point about using helpers instead of copying so much code. One potential helper function could check if the survey is already in tempdir?

The reason I ended up copying most of the `fetch_survey()` code is that it already does a really good job using helper functions for checking parameters, generating the URL, building the request, etc. Did you have some other sorts of functions in mind?

jmobrien commented:
Hey, just seeing this--this is great to have, thanks!  And nice catch on avoiding the saved files overwriting any from `fetch_survey()`

FYI, I just put in a PR adding features to fetch_survey() that also also tried to make the function a bit more modular (and restructured arg checking).  This includes a set of internal helper functions for the export-responses endpoint.  I *think* that could be useful as a drop-in for both `fetch_survey()` and `fetch_inprogress` that could help streamline things w/o too much effort. 

@juliasilge since these two functions are basically the same call, we could in theory consolidate just about everything into a shared internal function except the arg checks (if we're dropping the start/end args), payload/body build (same reason), and the file paths.  But I don't know how far you might want to go on DRY.

juliasilge commented:
@jamesmartherus Thank you again so much for this new feature! Are you interested in making the changes based on #263? I do think @jmobrien's idea of making one internal function that is used by both of these user-facing functions is the right way to go, for long-term maintainability and avoiding bugs down the road.

jmobrien commented:
There's a function called `export_responses_request()` that should work for that now.

juliasilge commented:
I just moved this repo's default branch from `master` to `main`; you can use [`usethis::git_default_branch_rediscover()`](https://usethis.r-lib.org/reference/git-default-branch.html) to update your local setup.
